### ISSUE TITLE
Locked in SharedBlockingCallback

### ISSUE THREAD

**Issue by openconcerto** [eclipse/jetty.project#1825]
Hi,

After seeing in logs :

2017-09-14 13:51:31.378:WARN:oejh.HttpParser:qtp1054932644-45: bad HTTP parsed: 400 for HttpChannelOverHttp@5b6c5174{r=0,c=false,a=IDLE,uri=null}
2017-09-14 13:51:31.379:WARN:oejh.HttpParser:qtp1054932644-40: Illegal character 0x16 in state=START for buffer HeapByteBuffer@58feb427[p=1,l=112,c=8192,r=111]={\x16<<<\x03\x01\x00k\x01\x00\x00g\x03\
x01Y\xBaw;MSc...\x13\x00\x12\x00\x11\x00\xFf\x01\x00\x00\x04\x00#\x00\x00>>>\x80\x00\x00\xFf\xDc\xE5\x9f\xC9\x15\xBa\xF7\xF286\xB9H\xC9...\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00}
2017-09-14 13:51:31.380:WARN:oejh.HttpParser:qtp1054932644-40: bad HTTP parsed: 400 Illegal character 0x16 for HttpChannelOverHttp@5c1ba46c{r=0,c=false,a=IDLE,uri=null}

The webserver stops responding to requests. Handlers are blocked in _outputstream write()_
as you can see in the attached [thread dump](https://github.com/eclipse/jetty.project/files/1308469/dump.txt)

The server is used in standalone mode 

```
final HttpConfiguration http_config = new HttpConfiguration();
http_config.setSecureScheme("https");
http_config.setSecurePort(port);
http_config.setRequestHeaderSize(8192);
http_config.setResponseHeaderSize(8192);
http_config.setSendServerVersion(false);
http_config.setSendDateHeader(false);
http_config.setIdleTimeout(10 * 60 * 1000); // 10 mins

final Server server = new Server(new QueuedThreadPool(16, 256));

 // SSL HTTP Configuration
final HttpConfiguration https_config = new HttpConfiguration(http_config);
https_config.addCustomizer(new SecureRequestCustomizer());

// SSL Connector
final ServerConnector sslConnector = new ServerConnector(server, new SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.asString()), new HttpConnectionFactory(https_config));
sslConnector.setPort(port);

ServerConnector httpConnector = new ServerConnector(server);
httpConnector.addConnectionFactory(new HttpConnectionFactory(http_config));
httpConnector.setPort(80);
server.setConnectors(new Connector[] { httpConnector, sslConnector });
```

The PublicResourceHandler used (blocked in write) is attached to a GzipHandler.

The server is running Ubuntu 16.04.2 LTS
and OpenJDK Runtime Environment (build 1.8.0_131-8u131-b11-2ubuntu1.16.04.3-b11)
( OpenJDK 64-Bit Server VM (build 25.131-b11, mixed mode) )

This issue have been seen more than 6 times in 2 months. 
Nothing except reported HttpParser issues seems related to such "dead server" situation.

I was not able to reproduce this bug on a test machine.

Best regards,
Guillaume

**Comment by joakime** [comment]
OP reports that occurs in Jetty 9.3.20 (was originally stated in title of issue)

**Comment by gregw** [comment]
The forever waiting in SharedBlockingCallback was an issue that we struggled with in 9.3.x and early 9.4.x releases.    Firstly, I do believe that we have covered all the edge cases in 9.4.x, so I would encourage an upgrade.

But if you are stuck on 9.3, then the issue occurs when some unusual event occurs on a connection and a callback to an outstanding write operation is lost somehow (typically due to a race with an error close). The writer then waits forever.     The fix needed significant restructuring of the error handling in jetty and that is one of the fundamental changes in 9.4 - thus it is not really something that can be backported to 9.3 (without turning 9.3 into 9.4).

To deal with these edge cases and to allow the server to recover, we have implemented a timeout on the blocking operations.  You can use `HttpConfiguration.setBlockingTimeout` to apply a total timeout to blocking operations, which would trigger in this case and cleanup after the unusual close of the connection.  This can be configured with the `jetty.httpConfig.blockingTimeout` property, which is normally set in the server module ini file.

Note that the server should not stop responding to requests unless it is configured with a very small thread pool and this issues affects all the available threads.

Note that the bad message exception does appear to be valid, as 0x16 is never a valid HTTP character.

Without an easy reproduction, it is hard for us to work out exactly what is happening and to make 9.3.x more robust.  I suspect that even if we knew, we probably could not fix all the issues as there were some fundamental races in the error handling in 9.3.x that we believe are fixed in 9.4.  So configuring the timeout is probably the best strategy for 9.3

**Comment by openconcerto** [comment]
Thank you for the explanation.
We will switch to 9.4.x as soon as possible. The quickfix `HttpConfiguration.setBlockingTimeout` is now deployed during our transition phase.

Maybe you could in the next 9.3.21 release set a default blocking timeout. It will avoid lots of issues for people staying with the 9.3.x who are not aware of the the potential "premature death" of their servers .

Best regards,
Guillaume

**Comment by gregw** [comment]
Some users might consider an unrequested timeout to be a "premature death", so I don't think we will change the default configuration now.   The conditions that trigger this problem appear to be rather rare in the wild and fixed in later releases

**Comment by Elishtar** [comment]
Hi, can you explain situation with blockingTimeout, we have the same issue on 9.3 and use blockingTimeout as solution, now we planning migration to 9.4. In cod of 9.4 I see that deprecated blockingTimeout, and recommendation(Replaced by setMinResponseDataRate and setMinRequestDataRate). My question is can we just remove blockingTimeout or we should use new methods(in context current issue)

**Comment by joakime** [comment]
The option called `blockingTimeout` on `HttpConfiguration` is a historical configuration that has outlived its original purpose.
In fact, it's been entirely removed from Jetty 10.0.x.

The options you have are a valid IdleTimeout (connector / httpclient / websocket).
Or using the data rate options: setMinResponseDataRate / setMinRequestDataRate.

**Comment by Elishtar** [comment]
Does that mean if just remove `blockingTimeout` after migration on 9.4, I can't reproduce the current issue? (we already have `IdleTimeout` on 9.3 issue fixed only after add blockingTimeout)

**Comment by sbordet** [comment]
@Elishtar since this issue was closed, you may want to open a new one.
My guess is that `blockingTimeout` was hiding your real issue.
Please detail in the new issue exactly what was happening in 9.3, why `blockingTimeout` was "working" for you, and we will be able to tell how to convert all that to 9.4.

**Comment by Elishtar** [comment]
@sbordet Ok, thank you for your answer
